File: ios/HydraICloudKV.m
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(HydraICloudKV, NSObject)

RCT_EXTERN__BLOCKING_SYNCHRONOUS_METHOD(getString:(NSString *)key)

RCT_EXTERN__BLOCKING_SYNCHRONOUS_METHOD(getAllKeys)

RCT_EXTERN_METHOD(set:(NSString *)key value:(NSString *)value)

RCT_EXTERN_METHOD(remove:(NSString *)key)

@end

File: ios/HydraICloudKV.swift
import Foundation
import React

/// Native wrapper for NSUbiquitousKeyValueStore (iCloud Key-Value Storage).
///
/// All methods are synchronous — NSUbiquitousKeyValueStore reads/writes are
/// in-process and do not block on network I/O. iCloud syncs in the background.
@objc(HydraICloudKV)
class HydraICloudKV: NSObject {

  private let store = NSUbiquitousKeyValueStore.default

  @objc static func requiresMainQueueSetup() -> Bool {
    return false
  }

  @objc func set(_ key: String, value: String) {
    store.set(value, forKey: key)
    store.synchronize()
  }

  @objc func getString(_ key: String) -> String? {
    return store.string(forKey: key)
  }

  @objc func remove(_ key: String) {
    store.removeObject(forKey: key)
    store.synchronize()
  }

  @objc func getAllKeys() -> [String] {
    return Array(store.dictionaryRepresentation.keys)
  }
}
